Update `RemoteAddrRoutePredicateFactory` to optionally respect the `X-Forwarded-For` header. Addresses #155.

Fitzoh opened the pull request:
Doesn't play nice with yaml config yet, haven't figured out how to parse additional arguments from `Tuple` yet.

Fitzoh commented:
Need to look into security implications as well, this is probably a good starting point cloudfoundry/gorouter#179

spencergibb commented:
I don't know how to do it with tuple. Seems like it's all or nothing. Maybe `@ConfigurationProperties`.

Fitzoh commented:
I'll start digging back into this one, should probably add a link to #125 as well.

Fitzoh commented:
Did some rebasing to catch up with the transition from String IPs -> InetSocketAddress and made it easier to customize the way the client IP is discovered.

By default, the existing behavior of `exchange.getRequest().getRemoteAddress()` is used.
There are now two other options:
`extractFinalRemoteAddressFromXForwardedHeader` always takes the first entry in the `X-Forwarded-For`, which is easy, but vulnerable to spoofing as described in some of the earlier referenced issues.
`extractRemoteAddressFromXForwardedHeaderByMaxTrustedIndex` requires an index (`maxTrustedIndex`) which corresponds to the number of trusted layers of infrastructure running in front of your application.

For instance, if you're running with just NGINX in front of the gateway, the `maxTrustedIndex` would be 1.

If you're running in Cloud Foundry with behind HAPROXY, then the gorouter, the `maxTrustedIndex` would be two.

The index is 1 based to prevent a user accidentally introducing an off by 1 error and erring too far in the untrusted direction.

Future work might include a remoteAddressResolver which takes a list of trusted IP addresses and terminates at the first untrusted IP. This would provide more flexibility if the number of ingress hops is unknown, but less helpful if trusted ip's are unknown.


Regarding `PredicateSpec`, I'm tempted to leave off a convenience function to increase the odds of the user reading the documentation, and ensuring that they're aware of the security implications.

spencergibb commented on spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/handler/support/RoutePredicateFactoryUtils.java:
```diff
@@ -17,14 +17,25 @@
  * @author Spencer Gibb
+ * @author Andrew Fitzgerald
  */
 public class RoutePredicateFactoryUtils {
```
I don't think this is the right place for this. After looking at the class, I'm going to remove it and move the one static method to the single class where it is used.

spencergibb replied:
Maybe a class called `RemoteAddressResolver`.

spencergibb commented on spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/handler/support/RoutePredicateFactoryUtils.java:
```diff
@@ -34,4 +45,82 @@ public static void traceMatch(String prefix, Object desired, Object actual, bool
+	 * @param maxTrustedIndex correlates to the number of trusted proxies expected in front spring cloud gateway
+	 *                           (index starts at 1).
+	 */
+	public static InetSocketAddress extractRemoteAddressFromXForwardedHeaderByMaxTrustedIndex(ServerWebExchange exchange, int maxTrustedIndex) {
```
These could possible attached to an enum and therefore used in normal configuration of `RemoteAddrRoutePredicateFactory`

Fitzoh commented:
Rebased and moved things out of `RoutePredicateFactoryUtils`.

There's now an `RemoteAddressResolver` interface, and two impls, default which just gets the client IP from the request, and an `X-Forwarded-For` based version.

The `X-Forwarded-For` uses the same logic around trusted number of infrastructure hops, and has two static constructor methods.
One takes a trusted index, and the other is a no-arg trust-all constructor...

Might need another round of polish, but wanted to check in after that rebase+refactor.
